src/property-filter/token-editor-grouped.tsx

```diff
@@ -0,0 +1,407 @@
+      setTemporaryToken({ ...temporaryToken, value: newValue });
+    };
+
+    return { token: temporaryToken, property, onChangePropertyKey, operator, onChangeOperator, value, onChangeValue };
+  });
+  return (
+    <div className={styles['token-editor-grouped']}>
+      <TokenEditorFields
+        supportsGroups={supportsGroups}
+        tokens={groups.map(group => getFormattedToken(group.token, i18nStrings))}
+        onRemove={index => {
+          const copy = [...tempGroup];
+          copy.splice(index, 1);
+          onChangeTempGroup(copy);
+        }}
+        onRemoveFromGroup={index => {
+          const removedToken = tempGroup[index];
+          const copy = [...tempGroup];
+          copy.splice(index, 1);
+          onChangeTempGroup(copy);
+          onChangeStandalone([...standaloneTokens, removedToken]);
+        }}
+        renderProperty={index => (
+          <PropertyInput
+            property={groups[index].property}
+            onChangePropertyKey={groups[index].onChangePropertyKey}
+            asyncProps={asyncProperties ? asyncProps : null}
+            filteringProperties={filteringProperties}
+            onLoadItems={onLoadItems}
+            customGroupsText={customGroupsText}
+            i18nStrings={i18nStrings}
+            freeTextFiltering={freeTextFiltering}
+          />
+        )}
+        renderOperator={index => (
+          <OperatorInput
+            property={groups[index].property}
+            operator={groups[index].operator}
+            onChangeOperator={groups[index].onChangeOperator}
+            i18nStrings={i18nStrings}
+            freeTextFiltering={freeTextFiltering}
+            triggerVariant="label"
+          />
+        )}
+        renderValue={index => (
+          <ValueInput
+            property={groups[index].property}
+            operator={groups[index].operator}
+            value={groups[index].value}
+            onChangeValue={groups[index].onChangeValue}
+            asyncProps={asyncProps}
+            filteringOptions={filteringOptions}
+            onLoadItems={onLoadItems}
+            i18nStrings={i18nStrings}
+          />
+        )}
+        i18nStrings={i18nStrings}
+      />
+
+      {supportsGroups && (
+        <div
+          className={clsx(styles['token-editor-grouped-add-token'], testUtilStyles['token-editor-token-add-actions'])}
+        >
+          <InternalButtonDropdown
+            variant="normal"
+            ariaLabel={i18nStrings.tokenEditorAddTokenActionsLabel}
+            items={standaloneTokens.map((token, index) => ({
+              id: index.toString(),
+              text: i18nStrings.tokenEditorAddExistingTokenLabel(getFormattedToken(token, i18nStrings)),
+            }))}
+            onItemClick={({ detail }) => {
+              const index = parseInt(detail.id);
+              const addedToken = standaloneTokens[index];
+              const copy = [...standaloneTokens];
+              copy.splice(index, 1);
```

Comment:
wouldn't it be better to use immutable methods for that? `.filter` is a good fit

Reply:
I personally find the .splice approach better explicit but don't mind either. Updated to .filter in both handlers